pkg/trace/api/evp_intake_proxy.go

```diff
@@ -0,0 +1,234 @@
+	if !r.conf.EVPProxy.Enabled {
+		return evpProxyErrorHandler("Has been disabled in config")
+	}
+	endpoints := evpProxyEndpointsFromConfig(r.conf)
+	transport := r.conf.NewHTTPTransport()
+	logger := stdlog.New(log.NewThrottled(5, 10*time.Second), "EVPProxy: ", 0) // limit to 5 messages every 10 seconds
+	handler := evpProxyForwarder(r.conf, endpoints, transport, logger)
+	return http.StripPrefix("/evp_proxy/v1", handler)
+}
+
+// evpProxyErrorHandler returns an HTTP handler that will always return
+// http.StatusMethodNotAllowed along with a clarification.
+func evpProxyErrorHandler(message string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		msg := fmt.Sprintf("EVPProxy is disabled: %v", message)
+		http.Error(w, msg, http.StatusMethodNotAllowed)
+	})
+}
+
+// evpProxyForwarder creates an http.ReverseProxy which can forward payloads to
+// one or more endpoints, based on the request received and the Agent configuration.
+// Headers are not proxied, instead we add our own known set of headers.
+// See also evpProxyTransport below.
+func evpProxyForwarder(conf *config.AgentConfig, endpoints []config.Endpoint, transport http.RoundTripper, logger *stdlog.Logger) http.Handler {
+	return &httputil.ReverseProxy{
+		Director: func(req *http.Request) {
+			req.Header["X-Forwarded-For"] = nil // Prevent setting X-Forwarded-For
+		},
+		ErrorLog:  logger,
+		Transport: &evpProxyTransport{transport, endpoints, conf},
+	}
+}
+
+// evpProxyTransport sends HTTPS requests to multiple targets using an
+// underlying http.RoundTripper. API keys are set separately for each target.
+// When multiple endpoints are in use the response from the first endpoint
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded.
+type evpProxyTransport struct {
+	transport http.RoundTripper
+	endpoints []config.Endpoint
+	conf      *config.AgentConfig
+}
+
+func (t *evpProxyTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
+	if req.Body != nil && t.conf.EVPProxy.MaxPayloadSize > 0 {
+		req.Body = apiutil.NewLimitedReader(req.Body, t.conf.EVPProxy.MaxPayloadSize)
+	}
+
+	// Metrics with stats for debugging
+	beginTime := time.Now()
+	metricTags := []string{}
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		metricTags = append(metricTags, "content_type:"+ct)
+	}
+	defer func() {
+		metrics.Count("datadog.trace_agent.evp_proxy.request", 1, metricTags, 1)
+		metrics.Count("datadog.trace_agent.evp_proxy.request_bytes", req.ContentLength, metricTags, 1)
+		metrics.Timing("datadog.trace_agent.evp_proxy.request_duration_ms", time.Since(beginTime), metricTags, 1)
+		if rerr != nil {
+			metrics.Count("datadog.trace_agent.evp_proxy.request_error", 1, metricTags, 1)
+		}
+	}()
+
+	// Read the input headers
+	subdomain := req.Header.Get("X-Datadog-EVP-Subdomain")
+	containerID := req.Header.Get("Datadog-Container-ID")
```

Comment:
```suggestion
	containerID := req.Header.Get(headerContainerID)
```

Reply:
Yes, I have this in mind. Depending on which one we merge first we should then update the other PR.

Reply:
Thanks!